packages/replay-next/components/elements-new/suspense/DOMReactCache.ts

```diff
@@ -0,0 +1,105 @@
+import { ObjectId, PauseId } from "@replayio/protocol";
+import { createCache } from "suspense";
+
+import { objectCache } from "replay-next/src/suspense/ObjectPreviews";
+import { pauseEvaluationsCache } from "replay-next/src/suspense/PauseCache";
+import { findProtocolObjectPropertyValue } from "replay-next/src/utils/protocol";
+import { ReplayClientInterface } from "shared/client/types";
+
+declare var __RECORD_REPLAY__: {
+  getObjectFromProtocolId?: (id: string | number) => any;
+};
+
+declare var __RECORD_REPLAY_ARGUMENTS__: {
+  getPersistentId?: (value: any) => number | null | void;
+};
+
+// TODO [FE-2005][FE-2067] With persistent DOM ids, we could switch this cache from PauseId to execution point.
+
+export type Data = [fiberId: number | null, displayName: string | null];
+
+export const domReactCache = createCache<
+  [replayClient: ReplayClientInterface, pauseId: PauseId, domNodeId: ObjectId],
+  Data
+>({
+  config: { immutable: true },
+  debugLabel: "DOMReactCache",
+  getKey: ([replayClient, pauseId, domNodeId]) => `${pauseId}:${domNodeId}`,
+  load: async ([replayClient, pauseId, domNodeId]) => {
+    const expression = `
+      {
+        const [id, displayName] = (${getReactInfo.toString()})(${domNodeId});
+        JSON.stringify([id, displayName]);
+      }
+    `;
+
+    const result = await pauseEvaluationsCache.readAsync(replayClient, pauseId, null, expression);
+    const value = result?.returned?.value;
+    if (value) {
+      return JSON.parse(value) as Data;
+    }
+
+    // This routine should always return a string value;
+    // if it returns an Object then that is likely an Error.
+    if (result.returned?.object) {
+      const errorObject = await objectCache.readAsync(
+        replayClient,
+        pauseId,
+        result.returned?.object,
+        "full"
+      );
+      const message = findProtocolObjectPropertyValue<string>(errorObject, "message");
+      if (message) {
+        console.error(message);
+      }
+    }
+
+    return [null, null];
+  },
+});
+
+function getReactInfo(domNodeId: ObjectId): Data {
+  if (
+    typeof __RECORD_REPLAY__ !== "undefined" &&
+    __RECORD_REPLAY__ != null &&
+    __RECORD_REPLAY__.getObjectFromProtocolId
+  ) {
+    const domNode = __RECORD_REPLAY__.getObjectFromProtocolId(domNodeId);
+
+    for (let name in domNode) {
+      // HACK
+      // Although a lot of tooling uses this attribute, it's not meant to be a public API.
+      // We already have a better DOM-node-to-Fiber-id mapping (coming from React DevTools)
+      // but it might not have been instantiated yet and it's in a place the Elements panel can't access.
+      // This approach is a work around with limited perf impact, but it's pretty hacky.
+      if (name.startsWith("__reactFiber$")) {
```

Comment:
Thanks for the comment. It's nice to see that we can leverage the data coming from React DevTools in the future.

Reply:
Note that the "React Event Listeners" functionality (both the J2C implementation and the actual "Event Listeners" panel) rely on this same implementation detail of React stuffing data related to the fiber onto the DOM node.